File: src/main/java/com/jolvera/jbud/HelloViewModel.java
package com.jolvera.jbud;

import com.jolvera.jbud.models.Transaction;
import javafx.beans.property.SimpleObjectProperty;
import javafx.beans.property.SimpleStringProperty;
import javafx.beans.property.StringProperty;
import javafx.collections.FXCollections;
import javafx.collections.ObservableList;

public class HelloViewModel {
    private final StringProperty transactionName = new SimpleStringProperty();
    private final StringProperty amount = new SimpleStringProperty("0.00");
    private final StringProperty status = new SimpleStringProperty();

    private final SimpleObjectProperty<ObservableList<Transaction>> transactionListProperty =
            new SimpleObjectProperty<>(FXCollections.observableArrayList());

    public HelloViewModel() {
        populateList();
    }

    private void populateList() {
        transactionListProperty.set(FXCollections.observableArrayList(
                new Transaction("Movies", 23.25),
                new Transaction("Groceries", 200.24),
                new Transaction("Makeup", 5.25)
        ));
    }

    public SimpleObjectProperty<ObservableList<Transaction>> getTransactionListProperty() {
        return transactionListProperty;
    }

    public StringProperty getTransactionName() {
        return transactionName;
    }

    public StringProperty getAmount() {
        return amount;
    }

    public StringProperty getStatus() {
        return status;
    }

    public void submitTransaction() {
        Transaction newTransaction = new Transaction(
                transactionName.get(),
                Double.parseDouble(amount.get())
        );

        var list = transactionListProperty.get();
        list.add(newTransaction);
        clear();
    }

    public void clear() {
        transactionName.set("");
        amount.set("");
    }
}

File: src/main/kotlin/com/jolvera/jbud/examples/HelloViewModel.kt
package com.jolvera.jbud.examples

import javafx.beans.property.*
import javafx.collections.FXCollections

class HelloViewModel {
    private val transactionNameProperty = SimpleStringProperty()
    private val amountProperty = SimpleStringProperty()
    private val transactionListProperty = SimpleObjectProperty(FXCollections.observableArrayList<Transaction>())

    init {
        populateList()
    }

    fun getTransactionNameProperty(): StringProperty = transactionNameProperty
    fun getAmountProperty(): StringProperty = amountProperty
    fun getTransactionListProperty() = transactionListProperty

    fun submitTransaction() {
        var newTransaction = Transaction(
            transactionNameProperty.get(),
            amountProperty.get().toDouble()
        )

        val list = transactionListProperty.get()
        list.add(newTransaction)
        clear();
    }

    private fun clear() {
        transactionNameProperty.set("")
        amountProperty.set("")
    }

    private fun populateList() {
        transactionListProperty.set(
            FXCollections.observableArrayList(
                Transaction("Movies", 24.50),
                Transaction("Groceries", 145.65),
                Transaction("Shovel", 12.45)
            )
        )
    }
}